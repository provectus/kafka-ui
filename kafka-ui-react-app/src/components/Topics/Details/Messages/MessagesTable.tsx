import React from 'react';
import { TopicMessage } from 'generated-sources';
import CustomParamButton from 'components/Topics/shared/Form/CustomParams/CustomParamButton';
import MessageItem from './MessageItem';

interface MessagesTableProp {
  messages: TopicMessage[];
  onNext(event: React.MouseEvent<HTMLButtonElement>): void;
}

const MessagesTable: React.FC<MessagesTableProp> = ({ messages, onNext }) => {
  if (!messages.length) {
    return <div>No messages at selected topic</div>;
  }

  return (
    <div>
      <table className="table is-striped is-fullwidth">
        <thead>
          <tr>
            <th>Timestamp</th>
            <th>Offset</th>
            <th>Partition</th>
            <th>Content</th>
          </tr>
        </thead>
        <tbody>
          {messages.map(
            ({ partition, offset, timestamp, content }: TopicMessage) => (
              <MessageItem
                key={timestamp.toString()}
                partition={partition}
                offset={offset}
                timestamp={timestamp}
                content={content as Record<string, unknown>}
              />
            )
          )}
        </tbody>
      </table>
      <div className="columns">
        <div className="column is-full">
          <CustomParamButton
            className="is-link is-pulled-right"
            type="fa-chevron-right"
            onClick={onNext}
            btnText="Next"
          />
        </div>
      </div>
    </div>
  );
};

export default MessagesTable;
