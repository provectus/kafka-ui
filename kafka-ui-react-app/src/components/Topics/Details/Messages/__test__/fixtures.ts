import { TopicMessage } from 'generated-sources';

export const messages: TopicMessage[] = [
  {
    partition: 1,
    offset: 2,
    timestamp: new Date('05-05-1995'),
    content: {
      foo: 'bar',
      key: 'val',
    },
  },
  {
    partition: 2,
    offset: 20,
    timestamp: new Date('05-07-2020'),
    content: undefined,
  },
];
