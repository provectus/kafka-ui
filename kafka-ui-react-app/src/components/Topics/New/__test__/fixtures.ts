import { CleanUpPolicy, Topic } from 'generated-sources';

export const createTopicPayload: Record<string, unknown> = {
  name: 'test-topic',
  partitions: 1,
  replicationFactor: 1,
  configs: {
    'cleanup.policy': 'delete',
    'retention.ms': '604800000',
    'retention.bytes': '-1',
    'max.message.bytes': '1000012',
    'min.insync.replicas': '1',
  },
};

export const createTopicResponsePayload: Topic = {
  name: 'local',
  internal: false,
  partitionCount: 1,
  replicationFactor: 1,
  replicas: 1,
  inSyncReplicas: 1,
  segmentSize: 0,
  segmentCount: 0,
  underReplicatedPartitions: 0,
  cleanUpPolicy: CleanUpPolicy.DELETE,
  partitions: [
    {
      partition: 0,
      leader: 1,
      replicas: [{ broker: 1, leader: false, inSync: true }],
      offsetMax: 0,
      offsetMin: 0,
    },
  ],
};
