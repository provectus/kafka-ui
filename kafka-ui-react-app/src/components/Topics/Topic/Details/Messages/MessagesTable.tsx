import PageLoader from 'components/common/PageLoader/PageLoader';
import { Table } from 'components/common/table/Table/Table.styled';
import TableHeaderCell from 'components/common/table/TableHeaderCell/TableHeaderCell';
import { TopicMessage } from 'generated-sources';
import React, { useContext } from 'react';
import {
  getTopicMessges,
  getIsTopicMessagesFetching,
} from 'redux/reducers/topicMessages/selectors';
import TopicMessagesContext from 'components/contexts/TopicMessagesContext';
import { useAppSelector } from 'lib/hooks/redux';

import Message from './Message';

const MessagesTable: React.FC = () => {
  const { isLive } = useContext(TopicMessagesContext);

  const messages = useAppSelector(getTopicMessges);
  const isFetching = useAppSelector(getIsTopicMessagesFetching);

  return (
    <Table isFullwidth>
      <thead>
        <tr>
          <TableHeaderCell> </TableHeaderCell>
          <TableHeaderCell title="Offset" />
          <TableHeaderCell title="Partition" />
          <TableHeaderCell title="Timestamp" />
          <TableHeaderCell title="Key" />
          <TableHeaderCell title="Content" />
          <TableHeaderCell> </TableHeaderCell>
        </tr>
      </thead>
      <tbody>
        {messages.map((message: TopicMessage) => (
          <Message
            key={[
              message.offset,
              message.timestamp,
              message.key,
              message.partition,
            ].join('-')}
            message={message}
          />
        ))}
        {isFetching && isLive && !messages.length && (
          <tr>
            <td colSpan={10}>
              <PageLoader />
            </td>
          </tr>
        )}
        {messages.length === 0 && !isFetching && (
          <tr>
            <td colSpan={10}>No messages found</td>
          </tr>
        )}
      </tbody>
    </Table>
  );
};

export default MessagesTable;
