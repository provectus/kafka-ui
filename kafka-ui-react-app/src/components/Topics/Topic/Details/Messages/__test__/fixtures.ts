import { TopicMessage } from 'generated-sources';

export const messages: TopicMessage[] = [
  {
    partition: 1,
    offset: 2,
    timestamp: new Date(Date.UTC(1995, 5, 5)),
    content: {
      foo: 'bar',
      key: 'val',
    },
  },
  {
    partition: 2,
    offset: 20,
    timestamp: new Date(Date.UTC(2020, 7, 5)),
    content: undefined,
  },
];

export const messageContent = `{
  "_id": "609fab8aed527f514f4e648d",
  "name": "in nostrud",
  "desc": "Dolore nostrud commodo magna velit ut magna voluptate sint aute. Excepteur aute culpa culpa dolor ipsum. Tempor est ut officia tempor laborum consectetur.\r\nAmet officia eu veniam Lorem enim aliqua aute voluptate elit do sunt in magna occaecat. Nisi sit non est adipisicing adipisicing consequat duis duis tempor consequat deserunt ea quis ad. Veniam sunt culpa nostrud adipisicing cillum voluptate non est cupidatat. Eiusmod tempor officia irure et deserunt est ex laboris occaecat adipisicing occaecat in aliquip aliqua. Do laboris culpa cupidatat cillum non. Ullamco excepteur mollit voluptate anim in nisi anim elit culpa aute. Ad officia sunt proident ut ullamco officia ea fugiat culpa cillum et fugiat aliquip.\r\nAmet non labore anim in ipsum. Et Lorem velit dolor ipsum. Irure id proident excepteur aliquip deserunt id officia dolor deserunt amet in sint. Aute in nostrud nulla ut laboris Lorem commodo nulla ipsum. Aliqua nulla commodo Lorem labore magna esse proident id ea in pariatur consectetur sint Lorem.\r\nCupidatat deserunt mollit tempor aliqua. Fugiat ullamco magna pariatur quis nulla magna. Esse duis labore ipsum nisi ullamco qui aute duis duis amet est laborum adipisicing magna. Est aliquip quis qui do aliquip nisi elit tempor ex aliquip. Excepteur aliquip ea deserunt amet adipisicing voluptate eiusmod sit sint exercitation exercitation. Id labore amet mollit ex commodo. Proident ex adipisicing deserunt esse Lorem tempor laborum nostrud commodo incididunt ea id.\r\n",
  "semster": "spring19",
  "profile": "cs",
  "degree": "bachelor",
  "degreee": "master",
  "degreeeee": "bachelor"
}`;
