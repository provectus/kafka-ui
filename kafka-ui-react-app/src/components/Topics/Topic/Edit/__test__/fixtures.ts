import { CleanUpPolicy, ConfigSource, TopicConfig } from 'generated-sources';
import { TopicWithDetailedInfo } from 'redux/interfaces/topic';

export const clusterName = 'testCluster';
export const topicName = 'testTopic';

export const config: TopicConfig[] = [
  {
    name: 'compression.type',
    value: 'producer',
    defaultValue: 'producer',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'compression.type',
        value: 'producer',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'compression.type',
        value: 'producer',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.value.schema.validation',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'leader.replication.throttled.replicas',
    value: '',
    defaultValue: '',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'confluent.key.subject.name.strategy',
    value: 'io.confluent.kafka.serializers.subject.TopicNameStrategy',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'message.downconversion.enable',
    value: 'true',
    defaultValue: 'true',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.downconversion.enable',
        value: 'true',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'min.insync.replicas',
    value: '1',
    defaultValue: '1',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'min.insync.replicas',
        value: '1',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'min.insync.replicas',
        value: '1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'segment.jitter.ms',
    value: '0',
    defaultValue: '0',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'cleanup.policy',
    value: 'delete',
    defaultValue: 'delete',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'cleanup.policy',
        value: 'delete',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'log.cleanup.policy',
        value: 'delete',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'flush.ms',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'confluent.tier.local.hotset.ms',
    value: '86400000',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.local.hotset.ms',
        value: '86400000',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'follower.replication.throttled.replicas',
    value: '',
    defaultValue: '',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'confluent.tier.local.hotset.bytes',
    value: '-1',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.local.hotset.bytes',
        value: '-1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.value.subject.name.strategy',
    value: 'io.confluent.kafka.serializers.subject.TopicNameStrategy',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'segment.bytes',
    value: '1073741824',
    defaultValue: '1073741824',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.segment.bytes',
        value: '1073741824',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'retention.ms',
    value: '604800000',
    defaultValue: '604800000',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'retention.ms',
        value: '604800000',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
    ],
  },
  {
    name: 'flush.messages',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.flush.interval.messages',
        value: '9223372036854775807',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.tier.enable',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.enable',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.tier.segment.hotset.roll.min.bytes',
    value: '104857600',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.tier.segment.hotset.roll.min.bytes',
        value: '104857600',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.segment.speculative.prefetch.enable',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.segment.speculative.prefetch.enable',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'message.format.version',
    value: '2.7-IV2',
    defaultValue: '2.7-IV2',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.format.version',
        value: '2.7-IV2',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'max.compaction.lag.ms',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.cleaner.max.compaction.lag.ms',
        value: '9223372036854775807',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'file.delete.delay.ms',
    value: '60000',
    defaultValue: '60000',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.segment.delete.delay.ms',
        value: '60000',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'max.message.bytes',
    value: '1000012',
    defaultValue: '1000012',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'max.message.bytes',
        value: '1000012',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'message.max.bytes',
        value: '1048588',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'min.compaction.lag.ms',
    value: '0',
    defaultValue: '0',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.cleaner.min.compaction.lag.ms',
        value: '0',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'message.timestamp.type',
    value: 'CreateTime',
    defaultValue: 'CreateTime',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.timestamp.type',
        value: 'CreateTime',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'preallocate',
    value: 'false',
    defaultValue: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.preallocate',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.placement.constraints',
    value: '',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'min.cleanable.dirty.ratio',
    value: '0.5',
    defaultValue: '0.5',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.cleaner.min.cleanable.ratio',
        value: '0.5',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'index.interval.bytes',
    value: '4096',
    defaultValue: '4096',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.index.interval.bytes',
        value: '4096',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'unclean.leader.election.enable',
    value: 'false',
    defaultValue: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'unclean.leader.election.enable',
        value: 'false',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'retention.bytes',
    value: '-1',
    defaultValue: '-1',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'retention.bytes',
        value: '-1',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'log.retention.bytes',
        value: '-1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'delete.retention.ms',
    value: '86400001',
    defaultValue: '86400000',
    source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'delete.retention.ms',
        value: '86400001',
        source: ConfigSource.DYNAMIC_TOPIC_CONFIG,
      },
      {
        name: 'log.cleaner.delete.retention.ms',
        value: '86400000',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.prefer.tier.fetch.ms',
    value: '-1',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'confluent.prefer.tier.fetch.ms',
        value: '-1',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'confluent.key.schema.validation',
    value: 'false',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'segment.ms',
    value: '604800000',
    defaultValue: '604800000',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [],
  },
  {
    name: 'message.timestamp.difference.max.ms',
    value: '9223372036854775807',
    defaultValue: '9223372036854775807',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.message.timestamp.difference.max.ms',
        value: '9223372036854775807',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
  {
    name: 'segment.index.bytes',
    value: '10485760',
    defaultValue: '10485760',
    source: ConfigSource.DEFAULT_CONFIG,
    isSensitive: false,
    isReadOnly: false,
    synonyms: [
      {
        name: 'log.index.size.max.bytes',
        value: '10485760',
        source: ConfigSource.DEFAULT_CONFIG,
      },
    ],
  },
];

export const partitions = [
  {
    partition: 0,
    leader: 2,
    replicas: [
      {
        broker: 2,
        leader: false,
        inSync: true,
      },
    ],
    offsetMax: 0,
    offsetMin: 0,
  },
];

export const topicWithInfo: TopicWithDetailedInfo = {
  name: topicName,
  internal: false,
  partitionCount: 1,
  replicationFactor: 1,
  replicas: 1,
  inSyncReplicas: 1,
  segmentSize: 0,
  segmentCount: 1,
  underReplicatedPartitions: 0,
  cleanUpPolicy: CleanUpPolicy.DELETE,
  partitions,
  config,
};
