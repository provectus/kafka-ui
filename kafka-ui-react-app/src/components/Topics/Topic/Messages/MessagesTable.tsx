import PageLoader from 'components/common/PageLoader/PageLoader';
import { Table } from 'components/common/table/Table/Table.styled';
import TableHeaderCell from 'components/common/table/TableHeaderCell/TableHeaderCell';
import { TopicMessage } from 'generated-sources';
import React, { useContext, useState } from 'react';
import {
  getTopicMessges,
  getIsTopicMessagesFetching,
} from 'redux/reducers/topicMessages/selectors';
import TopicMessagesContext from 'components/contexts/TopicMessagesContext';
import { useAppSelector } from 'lib/hooks/redux';
import { Button } from 'components/common/Button/Button';
import { useSearchParams } from 'react-router-dom';
import { MESSAGES_PER_PAGE } from 'lib/constants';
import * as S from 'components/common/NewTable/Table.styled';

import PreviewModal from './PreviewModal';
import Message, { PreviewFilter } from './Message';

const MessagesTable: React.FC = () => {
  const [previewFor, setPreviewFor] = useState<string | null>(null);

  const [keyFilters, setKeyFilters] = useState<PreviewFilter[]>([]);
  const [contentFilters, setContentFilters] = useState<PreviewFilter[]>([]);

  const [searchParams, setSearchParams] = useSearchParams();
  const page = searchParams.get('page');
  const { isLive } = useContext(TopicMessagesContext);

  const messages = useAppSelector(getTopicMessges);
  const isFetching = useAppSelector(getIsTopicMessagesFetching);

  const isTailing = isLive && isFetching;

  // Pagination is disabled in live mode, also we don't want to show the button
  // if we are fetching the messages or if we are at the end of the topic
  const isPaginationDisabled = isTailing || isFetching;

  const isNextPageButtonDisabled =
    isPaginationDisabled || messages.length < Number(MESSAGES_PER_PAGE);
  const isPrevPageButtonDisabled =
    isPaginationDisabled || !Number(searchParams.get('page'));

  const handleNextPage = () => {
    searchParams.set('page', String(Number(page || 0) + 1));
    setSearchParams(searchParams);
  };

  const handlePrevPage = () => {
    searchParams.set('page', String(Number(page || 0) - 1));
    setSearchParams(searchParams);
  };

  return (
    <div style={{ position: 'relative' }}>
      {previewFor !== null && (
        <PreviewModal
          values={previewFor === 'key' ? keyFilters : contentFilters}
          toggleIsOpen={() => setPreviewFor(null)}
          setFilters={(payload: PreviewFilter[]) =>
            previewFor === 'key'
              ? setKeyFilters(payload)
              : setContentFilters(payload)
          }
        />
      )}
      <Table isFullwidth>
        <thead>
          <tr>
            <TableHeaderCell> </TableHeaderCell>
            <TableHeaderCell title="Offset" />
            <TableHeaderCell title="Partition" />
            <TableHeaderCell title="Timestamp" />
            <TableHeaderCell
              title="Key"
              previewText={`Preview ${
                keyFilters.length ? `(${keyFilters.length} selected)` : ''
              }`}
              onPreview={() => setPreviewFor('key')}
            />
            <TableHeaderCell
              title="Value"
              previewText={`Preview ${
                contentFilters.length
                  ? `(${contentFilters.length} selected)`
                  : ''
              }`}
              onPreview={() => setPreviewFor('content')}
            />
            <TableHeaderCell> </TableHeaderCell>
          </tr>
        </thead>
        <tbody>
          {messages.map((message: TopicMessage) => (
            <Message
              key={[
                message.offset,
                message.timestamp,
                message.key,
                message.partition,
              ].join('-')}
              message={message}
              keyFilters={keyFilters}
              contentFilters={contentFilters}
            />
          ))}
          {isFetching && isLive && !messages.length && (
            <tr>
              <td colSpan={10}>
                <PageLoader />
              </td>
            </tr>
          )}
          {messages.length === 0 && !isFetching && (
            <tr>
              <td colSpan={10}>No messages found</td>
            </tr>
          )}
        </tbody>
      </Table>
      <S.Pagination>
        <S.Pages>
          <Button
            buttonType="secondary"
            buttonSize="L"
            disabled={isPrevPageButtonDisabled}
            onClick={handlePrevPage}
          >
            ← Back
          </Button>
          <Button
            buttonType="secondary"
            buttonSize="L"
            disabled={isNextPageButtonDisabled}
            onClick={handleNextPage}
          >
            Next →
          </Button>
        </S.Pages>
      </S.Pagination>
    </div>
  );
};

export default MessagesTable;
