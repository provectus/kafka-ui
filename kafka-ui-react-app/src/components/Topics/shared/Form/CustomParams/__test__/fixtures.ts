export const defaultValues = {
  partitions: 1,
  replicationFactor: 1,
  minInSyncReplicas: 1,
  cleanupPolicy: 'delete',
  retentionBytes: -1,
  maxMessageBytes: 1000012,
  name: 'TestCustomParamEdit',
  customParams: [
    {
      name: 'delete.retention.ms',
      value: '86400001',
    },
  ],
};
