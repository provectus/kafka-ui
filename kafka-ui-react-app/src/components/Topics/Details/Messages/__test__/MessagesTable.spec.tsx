import React from 'react';
import { shallow } from 'enzyme';
import MessagesTable, {
  MessagesTableProp,
} from 'components/Topics/Details/Messages/MessagesTable';
import { messages } from './fixtures';

jest.mock('date-fns', () => ({
  format: () => `mocked date`,
}));

describe('MessagesTable', () => {
  const setupWrapper = (props: Partial<MessagesTableProp> = {}) => (
    <MessagesTable messages={[]} onNext={jest.fn()} {...props} />
  );

  describe('when topic is empty', () => {
    it('renders table row with JSONTree', () => {
      const wrapper = shallow(setupWrapper());
      expect(wrapper.exists('table')).toBeFalsy();
      expect(wrapper.exists('CustomParamButton')).toBeFalsy();
      expect(wrapper.text()).toEqual('No messages at selected topic');
    });

    it('matches snapshot', () => {
      expect(shallow(setupWrapper())).toMatchSnapshot();
    });
  });

  describe('when topic contains messages', () => {
    const onNext = jest.fn();
    const wrapper = shallow(setupWrapper({ messages, onNext }));

    it('renders table row without JSONTree', () => {
      expect(wrapper.exists('table')).toBeTruthy();
      expect(wrapper.exists('CustomParamButton')).toBeTruthy();
      expect(wrapper.find('MessageItem').length).toEqual(2);
    });

    it('handles CustomParamButton click', () => {
      wrapper.find('CustomParamButton').simulate('click');
      expect(onNext).toHaveBeenCalled();
    });

    it('matches snapshot', () => {
      expect(shallow(setupWrapper({ messages, onNext }))).toMatchSnapshot();
    });
  });
});
