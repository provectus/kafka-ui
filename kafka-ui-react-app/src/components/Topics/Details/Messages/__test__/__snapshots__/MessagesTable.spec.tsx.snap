// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessagesTable when topic contains messages matches snapshot 1`] = `
<Fragment>
  <table
    className="table is-striped is-fullwidth"
  >
    <thead>
      <tr>
        <th>
          Timestamp
        </th>
        <th>
          Offset
        </th>
        <th>
          Partition
        </th>
        <th>
          Content
        </th>
      </tr>
    </thead>
    <tbody>
      <MessageItem
        content={
          Object {
            "foo": "bar",
            "key": "val",
          }
        }
        key="Fri May 05 1995 00:00:00 GMT+0400 (Moscow Summer Time)"
        offset={2}
        partition={1}
        timestamp={1995-05-04T20:00:00.000Z}
      />
      <MessageItem
        key="Thu May 07 2020 00:00:00 GMT+0300 (Moscow Standard Time)"
        offset={20}
        partition={2}
        timestamp={2020-05-06T21:00:00.000Z}
      />
    </tbody>
  </table>
  <div
    className="columns"
  >
    <div
      className="column is-full"
    >
      <CustomParamButton
        btnText="Next"
        className="is-link is-pulled-right"
        onClick={[MockFunction]}
        type="fa-chevron-right"
      />
    </div>
  </div>
</Fragment>
`;

exports[`MessagesTable when topic is empty matches snapshot 1`] = `
<div>
  No messages at selected topic
</div>
`;
